shared/tests/utility/support/TestYamlConversion.cpp

```diff
@@ -0,0 +1,197 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2021 NUbots <nubots@nubots.net>
+ */
+
+
+#include <Eigen/Core>
+#include <catch.hpp>
+#include <string>
+#include <utility>
+#include <yaml-cpp/yaml.h>
+
+#include "utility/support/yaml_expression.hpp"
+
+
+using utility::support::Expression;
+using utility::support::resolve_expression;
+
+/**
+ * Unit tests for yaml_expression.cpp
+ *
+ * @author Liam Craft
+ */
+SCENARIO("yaml nodes can be converted to a given container type", "[utility][support][yaml_expression]") {
+    GIVEN("A yaml node containing a list of data") {
+        YAML::Node test_values = YAML::LoadFile("tests/yamlConversionTestVals.yaml");
+        // error threshold when comparing doubles
+        static constexpr double ERROR_THRESHOLD = 1e-6;
+        // ints
+        WHEN("A vector is declared using integers from the YAML node") {
+            std::vector<int> int_vector = resolve_expression<int>(test_values["ints"]);
+            // Check that the values are good(container size, data type....)
+            THEN("Container sizes should be equivalent") {
+                REQUIRE(test_values["ints"].size() == int_vector.size());
+            }
+            THEN("Values at each index should be equivalent") {
+                // Check values at each index
+                for (size_t i = 0; i < test_values["ints"].size(); i++) {
+                    REQUIRE(test_values["ints"][i].as<int>() == int_vector.at(i));
+                }
+            }
+        }
+
+        // float
+        WHEN("A vector is declared using float values from the YAML node") {
+            std::vector<float> float_vector = resolve_expression<float>(test_values["floats"]);
+            THEN("Container sizes should be equivalent") {
+                REQUIRE(test_values["floats"].size() == float_vector.size());
+            }
+            THEN("Values at each index should be equivalent") {
+                for (size_t i = 0; i < test_values["floats"].size(); i++) {
+                    REQUIRE(test_values["floats"][i].as<float>() == float_vector.at(i));
+                }
+            }
+        }
+
+        // double
+        WHEN("A vector is declared using doubles from the YAML node") {
+            std::vector<double> double_vector = resolve_expression<double>(test_values["doubles"]);
+            THEN("Container sizes should be equivalent") {
+                REQUIRE(test_values["doubles"].size() == double_vector.size());
+            }
+            THEN("Values at each index should be equivalent") {
+                for (size_t i = 0; i < test_values["doubles"].size(); i++) {
+                    Approx epsilon_result = Approx(double_vector.at(i)).epsilon(ERROR_THRESHOLD);
```

Comment:
We use the "approx" comparison for doubles, but not for floats. Given that we are reading the values directly, I feel like this is one of the very rare occasions when direct comparison of floating point types is ok - as in, I don't think we need to check with `Approx` (unless this fails without it, in which case yaml-cpp is doing something funky, and we should probably use the `approx` for the `float`s as well?).

Reply:
I didn't use approx originally but it starts failing with "real" test data. Pretty much anything with something other than 0's on the end. The floats don't seem to mind for some reason? I'll change it for consistency if you reckon it's needed.